### PR TITLE
Assign generic role to aside tag within the sectioning content elements

### PR DESCRIPTION
**Pull request by joone**
#### 47f7deca4cff8a33ff3881f7d2801030f87c1d03
<pre>
Assign generic role to aside tag within the sectioning content elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=235065">https://bugs.webkit.org/show_bug.cgi?id=235065</a>
<a href="https://rdar.apple.com/problem/87391915">rdar://problem/87391915</a>

Reviewed by Tyler Wilcock.

This change assigns a generic ARIA role to &lt;aside&gt; elements when they are nested
within &lt;aside&gt;, &lt;article&gt;, &lt;section&gt;, or &lt;nav&gt; elements, aligning with the spec
(<a href="https://w3c.github.io/html-aam/#el-aside).">https://w3c.github.io/html-aam/#el-aside).</a>
This follows the discussion in <a href="https://github.com/w3c/html-aam/issues/512">https://github.com/w3c/html-aam/issues/512</a>

Additionally, it introduces a check to return LandmarkComplementary when an &lt;aside&gt;
tag has an explicit accessible name.

* LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):

Canonical link: <a href="https://commits.webkit.org/270509@main">https://commits.webkit.org/270509@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/131b7bc303aba6292969f5b466fee6831d318c1f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25657 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4262 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26940 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27757 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23501 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25940 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6010 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1697 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23637 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25906 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3170 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22118 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28337 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2801 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23071 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29145 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23407 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23437 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27000 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2825 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1061 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4202 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/22820 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6157 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3271 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3141 "Built successfully") | | |

### CODE REVIEW

File: LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html [hunk 1]
```diff
@@ -23,8 +23,9 @@
 
 <!-- el-aside -->
 <nav>
-  <aside data-testname="el-aside-in-section-with-name" data-expectedrole="complementary" aria-label="x" class="ex">x</aside>
+  <aside data-testname="el-aside-in-section-with-name" data-expectedrole="generic" aria-label="x" class="ex">x</aside>
```
Comment (twilco) [inline LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html]: I think the spec expects `aside`s with accessible names to have a role of `complementary`, so changing the expected role to `generic` isn't correct (based on my reading).

https://w3c.github.io/html-aam/#el-aside

Might be worth addressing this piece in a separate patch (if possible, might not be?)

File: LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html [hunk 2]
```diff
@@ -23,8 +23,9 @@
...
   <aside data-testname="el-aside-in-section-without-name" class="ex-generic">x</aside>
+  <aside data-testname="el-aside-in-section-with-role" data-expectedrole="complementary" class="ex" role="complementary">x</aside>
```
Comment (twilco) [inline LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html:28]: cc @cookiecrook 

Does this new testcase seem right based on your knowledge of the spec?

**Review by twilco — COMMENTED** [review]
Thanks for working on this! Left some comments.

File: LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html (hunk 1 above)
Comment (joone) [reply LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html]: I've rechecked the spec and you're correct. I'll update my patch for the <aside> to have complementary role.

File: LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html (hunk 2 above)
Comment (joone) [reply LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html:28]: There was a discussion about this: https://bugs.chromium.org/p/chromium/issues/detail?id=1459657#c12

As I undertand, when the aside element explicitly has the role="complementary", it should retain this role regardless of whether it's nested within other elements.

Related spec: https://www.w3.org/TR/core-aam-1.2/#mapping_conflicts

Comment (twilco) [reply LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html:28]: OK, makes sense. Because this adds a testcase to a WPT, we'll need to either:

1. Use `Tools/Scripts/export-w3c-test-changes` to export the change to https://github.com/web-platform-tests/wpt (never used this script before unfortunately, so don't have a ton more guidance)
2. Make a [PR to the WPT](https://github.com/web-platform-tests/wpt) outside of this WebKit PR, then re-import the test via `Tools/Scripts/import-w3c-tests -t web-platform-tests/html-aam`

Comment (joone) [reply LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html:28]: I've created a PR to the WPT by running ```Tools/Scripts/export-w3c-test-changes -c -g HEAD -b 235065```
https://github.com/web-platform-tests/wpt/pull/43028

File: Source/WebCore/accessibility/AccessibilityNodeObject.cpp [hunk 3]
```diff
@@ -440,8 +440,21 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
         return AccessibilityRole::LandmarkMain;
     if (node()->hasTagName(navTag))
         return AccessibilityRole::LandmarkNavigation;
-    if (node()->hasTagName(asideTag))
+    // The aside element should not assume the complementary role when nested
+    // within the sectioning content elements.
+    // https://w3c.github.io/html-aam/#el-aside
+    if (node()->hasTagName(asideTag)) {
+        // Return LandmarkComplementary if the aside element explicitly has a role="complementary".
```
Comment (twilco) [inline Source/WebCore/accessibility/AccessibilityNodeObject.cpp]: I think this comment is unnecessary — the code does a good job describing the intention. What do you think about removing this?

File: Source/WebCore/accessibility/AccessibilityNodeObject.cpp [hunk 4]
```diff
@@ -440,8 +440,21 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
         return AccessibilityRole::LandmarkMain;
     if (node()->hasTagName(navTag))
         return AccessibilityRole::LandmarkNavigation;
-    if (node()->hasTagName(asideTag))
+    // The aside element should not assume the complementary role when nested
+    // within the sectioning content elements.
+    // https://w3c.github.io/html-aam/#el-aside
```
Comment (twilco) [inline Source/WebCore/accessibility/AccessibilityNodeObject.cpp]: This is a good comment — let's move it closer to the code that it describes (`if (isDescendantOfElementType({ asideTag, articleTag, sectionTag, navTag }))`)

File: Source/WebCore/accessibility/AccessibilityNodeObject.cpp [hunk 5]
```diff
@@ -440,8 +440,21 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
...
+    if (node()->hasTagName(asideTag)) {
+        // Return LandmarkComplementary if the aside element explicitly has a role="complementary".
+        if (equalLettersIgnoringASCIICase(getAttribute(roleAttr), "complementary"_s))
+            return AccessibilityRole::LandmarkComplementary;
+        if (isDescendantOfElementType({ asideTag, articleTag, sectionTag, navTag })) {
+            // Return LandmarkComplementary if the aside element has an accessible name.
+            if (hasAttribute(aria_labelAttr))
```
Comment (twilco) [inline Source/WebCore/accessibility/AccessibilityNodeObject.cpp]: There are a lot of ways something can have an accessible name — `aria-label`, `aria-labelledby`, `aria-description`, `aria-describedby`, more. There are other roles with this requirement, so it would be nice if we had this abstracted in a shared function (e.g. `hasExplicitAccessibleName`).

But I need to think more on that, and don't want to block this PR on it. So for now, maybe we can just check for the presence of any four of those attributes, `hasAttribute(aria_labelAttr) || hasAttribute(aria_labelledby) || ...`? What do you think?

**Review by twilco — COMMENTED** [review]
Almost there!

File: Source/WebCore/accessibility/AccessibilityNodeObject.cpp [hunk 6]
```diff
@@ -440,8 +440,21 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
         return AccessibilityRole::LandmarkMain;
     if (node()->hasTagName(navTag))
         return AccessibilityRole::LandmarkNavigation;
-    if (node()->hasTagName(asideTag))
+    // The aside element should not assume the complementary role when nested
+    // within the sectioning content elements.
+    // https://w3c.github.io/html-aam/#el-aside
+    if (node()->hasTagName(asideTag)) {
+        // Return LandmarkComplementary if the aside element explicitly has a role="complementary".
+        if (equalLettersIgnoringASCIICase(getAttribute(roleAttr), "complementary"_s))
```
Comment (twilco) [inline Source/WebCore/accessibility/AccessibilityNodeObject.cpp]: Using `ariaRoleAttribute() == AccessibilityRole::LandmarkComplementary` will likely be more robust here, since I believe that something like `role="   complementary"` (note the whitespace) is still valid.

File: Source/WebCore/accessibility/AccessibilityNodeObject.cpp (hunk 3 above)
Comment (joone) [reply Source/WebCore/accessibility/AccessibilityNodeObject.cpp]: Okay, I will remove it.

File: Source/WebCore/accessibility/AccessibilityNodeObject.cpp (hunk 4 above)
Comment (joone) [reply Source/WebCore/accessibility/AccessibilityNodeObject.cpp]: Sure!

File: Source/WebCore/accessibility/AccessibilityNodeObject.cpp (hunk 5 above)
Comment (joone) [reply Source/WebCore/accessibility/AccessibilityNodeObject.cpp]: Okay,  I will check other aria attributes here like:
```
  if (hasAttribute(aria_labelAttr) || hasAttribute(aria_labelledbyAttr) || hasAttribute(aria_descriptionAttr) || hasAttribute(aria_describedbyAttr))
```
Then, I'll start working on a separate patch to introduce hasExplicitAccessibleName().

File: Source/WebCore/accessibility/AccessibilityNodeObject.cpp (hunk 6 above)
Comment (joone) [reply Source/WebCore/accessibility/AccessibilityNodeObject.cpp]: Done.

**Review by twilco — APPROVED** [review]
Thank you @joone!
